src_zlib/service/zlib_service.cpp

```diff
@@ -262,111 +183,91 @@ struct service_impl
+        int level,
+        int window_bits,
+        int mem_level)
     {
-        return {
-            clamp(out.size()) - zs.avail_out,
-            clamp(in.size()) - zs.avail_in,
-            ret < 0 ? static_cast<error>(ret) : system::error_code{},
-            ret == Z_STREAM_END };
+        zs_.zalloc = &zalloc;
+        zs_.zfree  = &zfree;
+        zs_.opaque = &ws;
+
+        auto ret = deflateInit2(&zs_, level, Z_DEFLATED,
+            window_bits, mem_level, Z_DEFAULT_STRATEGY);
+        if(ret != Z_OK)
+            throw_zlib_error(ret);
     }
 
-    class deflator
-        : public stream
+    system::error_code
+    write(params& p, flush f) noexcept override
     {
-        z_stream zs_;
-
-    public:
-        deflator(
-            http_proto::detail::workspace& ws,
-            int level,
-            int window_bits,
-            int mem_level)
-        {
-            zs_.zalloc = &zalloc_impl;
-            zs_.zfree  = &zfree_impl;
-            zs_.opaque = &ws;
-
-            auto ret = deflateInit2(&zs_, level, Z_DEFLATED,
-                window_bits, mem_level, Z_DEFAULT_STRATEGY);
-            if(ret != Z_OK)
-                throw_zlib_error(ret);
-        }
-
-        virtual results
-        write(
-            buffers::mutable_buffer out,
-            buffers::const_buffer in,
-            flush flush) noexcept override
-        {
-            sync(&zs_, out, in);
-            return make_results(zs_, out, in,
-                deflate(&zs_, static_cast<int>(flush)));
-        }
-    };
-
-    class inflator
-        : public stream
+        auto adp = adapter{ &zs_, &p };
+        return static_cast<error>(
+            deflate(&zs_, static_cast<int>(f)));
+    }
+};
+
+class inflator
+    : public stream
+{
+    z_stream zs_;
+
+public:
+    inflator(
+        http_proto::detail::workspace& ws,
+        int window_bits)
     {
-        z_stream zs_;
-
-    public:
-        inflator(
-            http_proto::detail::workspace& ws,
-            int window_bits)
-        {
-            zs_.zalloc = &zalloc_impl;
-            zs_.zfree  = &zfree_impl;
-            zs_.opaque = &ws;
-
-            auto ret = inflateInit2(&zs_, window_bits);
-            if(ret != Z_OK)
-                throw_zlib_error(ret);
-        }
-
-        virtual results
-        write(
-            buffers::mutable_buffer out,
-            buffers::const_buffer in,
-            flush flush) noexcept override
-        {
-            sync(&zs_, out, in);
-            return make_results(zs_, out, in,
-                inflate(&zs_, static_cast<int>(flush)));
-        }
-    };
+        zs_.zalloc = &zalloc;
+        zs_.zfree  = &zfree;
+        zs_.opaque = &ws;
+
+        auto ret = inflateInit2(&zs_, window_bits);
+        if(ret != Z_OK)
+            throw_zlib_error(ret);
+    }
+
+    system::error_code
+    write(params& p, flush f) noexcept override
+    {
+        auto adp = adapter{ &zs_, &p };
```

Comment:
I would rather this be two explicit function calls instead of the destructor which hides things, but since it is an implementation detail I am not going to complain too much.

Reply:
I agree and I'll change it.